packages/next/src/server/app-render/use-flight-response.tsx

```diff
@@ -86,25 +49,124 @@ export function useFlightResponse(
     },
     nonce,
   })
-  flightResponses.set(flightStream, res)
 
-  pipeFlightDataToInlinedStream(forwardStream, writable, nonce, formState)
+  flightResponses.set(flightStream, newResponse)
 
-  return res
+  return newResponse
+}
+
+/**
+ * There are times when an SSR render may be finished but the RSC render
+ * is ongoing and we need to wait for it to complete to make some determination
+ * about how to handle the render. This function will drain the RSC reader and
+ * resolve when completed. This will generally require teeing the RSC stream and it
+ * should be noted that it will cause all the RSC chunks to queue in the underlying
+ * ReadableStream however given Flight currently is a push stream that doesn't respond
+ * to backpressure this shouldn't change how much memory is maximally consumed
+ */
+export async function flightRenderComplete(
+  flightStream: ReadableStream<Uint8Array>
+): Promise<void> {
+  const flightReader = flightStream.getReader()
+
+  return new Promise(async (resolve, reject) => {
+    try {
+      while (true) {
+        const { done } = await flightReader.read()
+        if (done) {
+          resolve()
+          return
+        }
+      }
+    } catch (error) {
+      reject(error)
+    }
+  })
```

Comment:
Fundamentally, this would be the same as writing:

```suggestion
  while (true) {
    const { done } = await flightReader.read()
    if (done) {
      return
    }
  }
```

As the contents of the handler for the promise is executed sync, and the wrapping function is already marked as `async`.